x/slpp/keeper/keeper.go

```diff
@@ -4,15 +4,197 @@ import (
+			sb, types.UniqueIndexAVSKeyPrefix, "avs_id_unique_idx", collections.Uint64Key, collections.StringKey,
+			func(_ string, avs types.AVS) (uint64, error) {
+				return avs.Id, nil
+			},
+		),
+		idMulti: indexes.NewMulti[uint64, string, types.AVS](
+			sb, types.IDIndexAVSKeyPrefix, "avs_id_idx", collections.Uint64Key, collections.StringKey,
+			func(_ string, avs types.AVS) (uint64, error) {
+				return avs.Id, nil
+			},
+		),
+	}
+}
+
+type WasmMsgServer interface {
+	StoreCode(ctx context.Context, msg *wasmtypes.MsgStoreCode) (*wasmtypes.MsgStoreCodeResponse, error)
+	InstantiateContract(ctx context.Context, msg *wasmtypes.MsgInstantiateContract) (*wasmtypes.MsgInstantiateContractResponse, error)
+}
+
+// Keeper is the base keeper for the x/oracle module.
+type Keeper struct {
+	storeService store.KVStoreService
+	cdc          codec.BinaryCodec
+
+	wasmMsgServer WasmMsgServer
+
+	// schema
+	nextAVSID collections.Sequence
+	avsMap    *collections.IndexedMap[string, types.AVS, *avsIndices]
+	schema    collections.Schema
+
+	// indexes
+	idIndex *indexes.Multi[uint64, string, types.AVS]
+}
+
+func NewKeeper(
+	ss store.KVStoreService,
+	cdc codec.BinaryCodec,
+	wasmMsgServer WasmMsgServer,
+) Keeper {
+	// create a new schema builder
+	sb := collections.NewSchemaBuilder(ss)
+
+	indices := newAVSIndices(sb)
+
+	idMulti, ok := indices.IndexesList()[1].(*indexes.Multi[uint64, string, types.AVS])
+	if !ok {
+		panic("expected idMulti to be a *indexes.Multi[uint64, string, types.AVS]")
+	}
+
+	k := Keeper{
+		storeService:  ss,
+		cdc:           cdc,
+		nextAVSID:     collections.NewSequence(sb, types.AVSIDKeyPrefix, "avs_id"),
+		avsMap:        collections.NewIndexedMap(sb, types.AVSKeyPrefix, "avs", collections.StringKey, codec.CollValue[types.AVS](cdc), indices),
+		idIndex:       idMulti,
+		wasmMsgServer: wasmMsgServer,
+	}
+
+	// create the schema
+	schema, err := sb.Build()
+	if err != nil {
+		panic(err)
+	}
+
+	k.schema = schema
+	return k
+}
+
+func (k *Keeper) HasAVSContract(ctx sdk.Context, contractBytes []byte) bool {
+	//calculate sha256 of contract bytes
+	ok, err := k.avsMap.Has(ctx, hex.EncodeToString(crypto.Sha256(contractBytes)))
+	if err != nil || !ok {
+		return false
+	}
+
+	return true
+}
+
+func (k *Keeper) NextAVSID(ctx sdk.Context) (uint64, error) {
+	return k.nextAVSID.Peek(ctx)
+}
+
+func (k *Keeper) RegisterAVS(ctx sdk.Context, m *types.MsgRegisterAVS) (uint64, error) {
+	if k.HasAVSContract(ctx, m.GetContractBin()) {
+		return 0, types.NewAVSContractAlreadyExistsError(hex.EncodeToString(m.GetContractBin()))
+	}
+
+	id, err := k.nextAVSID.Next(ctx)
+	if err != nil {
+		return 0, err
+	}
+
+	storeCodeResponse, err := k.wasmMsgServer.StoreCode(ctx, &wasmtypes.MsgStoreCode{
+		Sender:       m.Sender,
+		WASMByteCode: m.ContractBin,
+	})
+	instantiateContractResponse, err := k.wasmMsgServer.InstantiateContract(ctx, &wasmtypes.MsgInstantiateContract{
+		Sender: m.Sender,
+		Admin:  m.Sender,
```

Comment:
nit (not relevant for hackathon, but would be my real comment): We should not set the admin as the msg sender, cause it then allows the original caller to always migrate and do something diffferent without a process. This should be optional so we can just remove (and then in the future if we wanted the ability to change the avs contract we should open up another msg type in SLPP that allows us to do so with extra checks